blockwork/build/transform.py

```diff
@@ -13,22 +13,48 @@
 # limitations under the License.
 
 from typing import TYPE_CHECKING, Any, Iterable
+
+from ..common.inithooks import InitHooks
 from ..build.interface import Interface, Direction
 if TYPE_CHECKING:
     from ..tools.tool import Version, Tool, Invocation
     from ..context import Context
 from ..common.complexnamespaces import ReadonlyNamespace
 
+@InitHooks()
 class Transform:
     """
     Base class for transforms.
     """
     tools: list[type["Tool"]] = []
-
-    def __init__(self):
-        self.input_interfaces: list[Interface] = []
-        self.output_interfaces: list[Interface] = []
-        self.interfaces_by_name: dict[str, tuple[Direction, Interface]] = {}
+    _interfaces: dict[str, tuple[Direction, Interface]] 
+
+    @InitHooks.pre
+    def init_interfaces(self):
+        self._interfaces = {}
+
+    @property
+    def output_interfaces(self):
+        interfaces: dict[str, Any] = {}
+        for name, (direction, interface) in self._interfaces.items():
+            if direction.is_output:
+                interfaces[name] = interface
+        return ReadonlyNamespace(**interfaces)
+
+    @property
+    def input_interfaces(self):
+        interfaces: dict[str, Any] = {}
+        for name, (direction, interface) in self._interfaces.items():
+            if direction.is_input:
+                interfaces[name] = interface
+        return ReadonlyNamespace(**interfaces)
```

Comment:
Are these calculations that could be cached? It feels like this might get expensive if these functions are called a lot

Reply:
Updated to cache these